File: solutions/206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
​
using Node = ListNode;
​
class Solution {
public:
    ListNode* reverseList(ListNode* root) {
        if(root == NULL or root->next == NULL){
            return root;
        }
        Node* nroot = reverseList(root->next);
        root->next->next = root;
        root->next = NULL;
        return nroot;
    }
};

File: ..solutions/206. Reverse Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;
        return iterativeReverse(head);
    }
    public ListNode iterativeReverse(ListNode node) {
        ListNode previous = null;
        ListNode current = node;
        while (current != null) {
            ListNode nextnode = current.next;
            current.next = previous;
            previous = current;
            current = nextnode;
        }
        return previous;
    }
    public ListNode recursiveReverse(ListNode node) {
        if (node.next==null)
            return node;

        ListNode newhead = recursiveReverse(node.next);
        node.next.next = node;
        node.next = null;

        return newhead;
    }
}
